File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.java
class TrieNode{
    Map<Character, TrieNode> children;
    boolean wordEnd;
    
    public TrieNode() {
        this.children = new HashMap<>();
        this.wordEnd = false;
    }
}

class Trie {
    TrieNode root;

    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode curr = root;
        
        for(int i =0;i<word.length();i++){
            char temp = word.charAt(i);
            if(!curr.children.containsKey(temp)){
                curr.children.put(temp, new TrieNode());
            }
            curr = curr.children.get(temp);
        }
        curr.wordEnd = true;
    }
    
    public boolean search(String word) {
        TrieNode curr = root;
        
        for(int i = 0; i<word.length(); i++){
            char temp = word.charAt(i);
            if(!curr.children.containsKey(temp)){
                return false;
            }
            curr = curr.children.get(temp);
        }
        return curr.wordEnd;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode curr = root;
        
        for(int i = 0; i<prefix.length(); i++){
            char temp = prefix.charAt(i);
            if(!curr.children.containsKey(temp)){
                return false;
            }
            curr = curr.children.get(temp);
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.py
class TrieNode:
    def __init__(self):
        # dict of character to TrieNode
        self.children = dict()
        self.wordEnd = False

class Trie:

    def __init__(self):
        self.root = TrieNode()
        

    def insert(self, word: str) -> None:
        curr = self.root
        for s in word:
            if s not in curr.children:
                curr.children[s] = TrieNode()
            curr = curr.children[s]
        curr.wordEnd = True
                

    def search(self, word: str) -> bool:
        curr = self.root
        for s in word:
            if s not in curr.children:
                return False
            curr = curr.children[s]
        return curr.wordEnd 
        

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for s in prefix:
            if s not in curr.children:
                return False
            curr = curr.children[s]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)